File: RPG/Item.cpp
#include "Item.h"
#include "Hero.h" 

Item::Item(ItemsType type) : type(type) {
    // Initialize item properties
    switch (type) {
    case ItemsType::POTION:
        effectValue = 10;
        description = "Restores 10 HP";
        break;
    case ItemsType::ATTACK_STONE:
        effectValue = 1;
        description = "+1 Attack permanently";
        break;
    case ItemsType::DEFENSE_STONE:
        effectValue = 2;
        description = "+2 Defense permanently";
        break;
    case ItemsType::ETHER:
        effectValue = 5;
        description = "Restores 5 MP";
        break;
    }
}

Item::~Item() {
    if (useSound.frameCount != 0) {
        UnloadSound(useSound);
    }
}

bool Item::UseItem(Hero& hero) {
    bool wasUsed = false;

    switch (type) {
    case ItemsType::POTION: {
        int maxHp = hero.GetMaxHp();
        int currentHp = hero.GetHp();

        if (currentHp >= maxHp) {
            TraceLog(LOG_WARNING, "HP already full!");
            wasUsed = false;
        }
        else {
            int newHp = std::min(maxHp, currentHp + effectValue);
            hero.setHp(newHp);
            TraceLog(LOG_INFO, "Healed %d HP!", effectValue);
            wasUsed = true;
        }
        break;
    }

    case ItemsType::ATTACK_STONE: {
        int currentAttack = hero.GetAttack();
        if (currentAttack >= 100) {
            TraceLog(LOG_WARNING, "Attack at maximum!");
            wasUsed = false;
        }
        else {
            hero.SetAttack(currentAttack + effectValue);
            TraceLog(LOG_INFO, "Attack increased by %d!", effectValue);
            wasUsed = true;
        }
        break;
    }

    case ItemsType::DEFENSE_STONE: {
        int currentDefense = hero.GetDefense();
        if (currentDefense >= 100) {
            TraceLog(LOG_WARNING, "Defense at maximum!");
            wasUsed = false;
        }
        else {
            hero.SetDefense(currentDefense + effectValue);
            TraceLog(LOG_INFO, "Defense increased by %d!", effectValue);
            wasUsed = true;
        }
        break;
    }

    case ItemsType::ETHER: {
        int maxMp = hero.GetMaxMp();
        int currentMp = hero.GetMp();

        if (currentMp >= maxMp) {
            TraceLog(LOG_WARNING, "MP already full!");
            wasUsed = false;
        }
        else {
            int newMp = std::min(maxMp, currentMp + effectValue);
            hero.SetMp(newMp);
            TraceLog(LOG_INFO, "Restored %d MP!", effectValue);
            wasUsed = true;
        }
        break;
    }
    }

    // Play sound if item was used successfully
    if (wasUsed && useSound.frameCount != 0) {
        PlaySound(useSound);
    }

    return wasUsed;
}

std::string Item::ToString() const {
    return description;
}

File: Item.java
package YushanoBoken.Model;

public class Item{
    private ItemsType type;
    private String description;
    private int effectOfItem;

    public Item(ItemsType type) {
        this.type = type;
        if (type == ItemsType.POTION){
            this.effectOfItem = 10;
            this.description= "Health 10 points of your Hp";
        } else if (type == ItemsType.ATTACK_STONE){
            this.effectOfItem = 1;
            this.description= "Increase your Attack Stat plus 1";
        } else if (type == ItemsType.DEFENSE_STONE){
            this.effectOfItem = 2;
            this.description= "Increase your Defense Stat plus 2";
        } else if (type == ItemsType.ETHER){
            this.effectOfItem = 5;
            this.description= "Restore your MP by 5";
        }
    }

    public ItemsType getType() {
        return type;
    }

    public boolean useItem(Hero hero){
        if(type.equals(ItemsType.POTION)){
            if (hero.hp >= hero.getMaxHeroHp()){
                System.out.println("You can't use it right now");
                return false;
            } else if (hero.hp <= hero.getMaxHeroHp()) {
                hero.setHp(hero.getHp() + effectOfItem);
                if(hero.hp > hero.getMaxHeroHp()){
                    hero.setHp(hero.getMaxHeroHp());
                }
                System.out.println("You have been healed");
                return true;
            }
        } else if (type.equals(ItemsType.ATTACK_STONE)){
            if(hero.attack == 100){
                System.out.println("You can no longer increase your attack ");
                return false;
            }else{
                hero.setAttack(hero.getAttack() + effectOfItem);
                System.out.println("You attack have been increased");
                return true;
            }
        } else if (type.equals(ItemsType.DEFENSE_STONE)) {
            if(hero.defense == 100){
                System.out.println("You can no longer increase your defense ");
                return false;
            }else {
                hero.setDefense(hero.getDefense() + effectOfItem);
                System.out.println("You defense have been increased");
                return true;
            }
        }else if(type.equals(ItemsType.ETHER)){
            if (hero.mp >= hero.getMaxHeroMp()){
                System.out.println("You can't use it right now");
                return false;
            }else if (hero.mp <= hero.getMaxHeroMp()) {
                hero.setMp(hero.getMp() + effectOfItem);
                if(hero.mp > hero.getMaxHeroMp()){
                    hero.setMp(hero.getMaxHeroMp());
                }
                System.out.println("You MP have been restored");
                return true;
            }

        }
        return false;
    }

    public String toString(){
        return type + " " + description;
    }

}